File: frontend/src/lib/nodeAvailability.ts
/**
 * Node Availability Helper
 *
 * Manages chatbot node availability based on WhatsApp connection type.
 */

export type ConnectionType = 'official' | 'qrcode';
export type NodeStatus = 'available' | 'experimental' | 'unavailable';

export interface NodeMetadata {
  available: boolean;
  status: NodeStatus;
  warning: string | null;
  connection_type: ConnectionType;
}

export interface NodeAvailabilityInfo {
  available_node_types: string[];
  node_metadata: Record<string, NodeMetadata>;
}

/**
 * Check if a node type is available for a connection type
 */
export function isNodeAvailable(
  nodeType: string,
  availabilityInfo: NodeAvailabilityInfo | null
): boolean {
  if (!availabilityInfo) return true; // Assume available if no info

  return availabilityInfo.available_node_types.includes(nodeType);
}

/**
 * Get node status (available, experimental, unavailable)
 */
export function getNodeStatus(
  nodeType: string,
  availabilityInfo: NodeAvailabilityInfo | null
): NodeStatus {
  if (!availabilityInfo) return 'available';

  const metadata = availabilityInfo.node_metadata[nodeType];
  return metadata?.status || 'available';
}

/**
 * Get warning message for a node type
 */
export function getNodeWarning(
  nodeType: string,
  availabilityInfo: NodeAvailabilityInfo | null
): string | null {
  if (!availabilityInfo) return null;

  const metadata = availabilityInfo.node_metadata[nodeType];
  return metadata?.warning || null;
}

/**
 * Get badge label for node status
 */
export function getNodeBadgeLabel(status: NodeStatus): string | null {
  switch (status) {
    case 'experimental':
      return 'Experimental';
    case 'unavailable':
      return 'Indisponível';
    default:
      return null;
  }
}

/**
 * Get badge color for node status (Tailwind classes)
 */
export function getNodeBadgeColor(status: NodeStatus): string {
  switch (status) {
    case 'experimental':
      return 'bg-yellow-100 text-yellow-800 border-yellow-300';
    case 'unavailable':
      return 'bg-red-100 text-red-800 border-red-300';
    default:
      return 'bg-green-100 text-green-800 border-green-300';
  }
}

/**
 * Filter available nodes from a list
 */
export function filterAvailableNodes(
  nodeTypes: string[],
  availabilityInfo: NodeAvailabilityInfo | null
): string[] {
  if (!availabilityInfo) return nodeTypes;

  return nodeTypes.filter(nodeType =>
    isNodeAvailable(nodeType, availabilityInfo)
  );
}

/**
 * Get all unavailable node types
 */
export function getUnavailableNodes(
  availabilityInfo: NodeAvailabilityInfo | null
): string[] {
  if (!availabilityInfo) return [];

  return Object.entries(availabilityInfo.node_metadata)
    .filter(([_, metadata]) => !metadata.available)
    .map(([nodeType, _]) => nodeType);
}

/**
 * Get all experimental node types
 */
export function getExperimentalNodes(
  availabilityInfo: NodeAvailabilityInfo | null
): string[] {
  if (!availabilityInfo) return [];

  return Object.entries(availabilityInfo.node_metadata)
    .filter(([_, metadata]) => metadata.status === 'experimental')
    .map(([nodeType, _]) => nodeType);
}

/**
 * Format node type for display
 */
export function formatNodeType(nodeType: string): string {
  const typeMap: Record<string, string> = {
    start: 'Start',
    message: 'Message',
    question: 'Question',
    condition: 'Condition',
    end: 'End',
    handoff: 'Handoff',
    delay: 'Delay',
    jump: 'Jump',
    action: 'Action',
    api_call: 'API Call',
    ai_prompt: 'AI Prompt',
    database_query: 'Database Query',
    script: 'Script',
    whatsapp_template: 'WhatsApp Template',
    interactive_buttons: 'Interactive Buttons',
    interactive_list: 'Interactive List',
  };

  return typeMap[nodeType] || nodeType;
}

File: backend/app/utils/node_availability.py
"""
Node availability based on WhatsApp connection type.

Defines which chatbot nodes are available for each WhatsApp connection type:
- official (Meta Cloud API)
- qrcode (Evolution API)
"""

from typing import List, Dict, Any


class NodeAvailability:
    """Manage node availability based on WhatsApp connection type."""

    # All available node types
    ALL_NODES = [
        "start",
        "message",
        "question",
        "condition",
        "end",
        "handoff",
        "delay",
        "jump",
        "action",
        "api_call",
        "ai_prompt",
        "database_query",
        "script",
        "set_variable",
        "random",
        "datetime",
        "analytics",
        "whatsapp_template",
        "interactive_buttons",
        "interactive_list",
    ]

    # Nodes that work on BOTH official and qrcode
    UNIVERSAL_NODES = [
        "start",
        "message",
        "question",
        "condition",
        "end",
        "handoff",
        "delay",
        "jump",
        "action",
        "api_call",
        "ai_prompt",
        "database_query",
        "script",
        "set_variable",
        "random",
        "datetime",
        "analytics",
    ]

    # Nodes EXCLUSIVE to Meta Cloud API (official)
    OFFICIAL_ONLY_NODES = [
        "whatsapp_template",  # Templates require Meta approval
    ]

    # Nodes that work but are EXPERIMENTAL on Evolution API
    EXPERIMENTAL_NODES = [
        "interactive_buttons",  # Works via Baileys (unofficial)
        "interactive_list",     # Works via Baileys (unofficial)
    ]

    @classmethod
    def get_available_nodes(cls, connection_type: str) -> List[str]:
        """
        Get list of available node types for a connection type.

        Args:
            connection_type: 'official' or 'qrcode'

        Returns:
            List of available node type strings
        """
        if connection_type == "official":
            # Official API supports ALL nodes
            return cls.ALL_NODES.copy()
        elif connection_type == "qrcode":
            # Evolution API supports universal + experimental nodes
            return cls.UNIVERSAL_NODES + cls.EXPERIMENTAL_NODES
        else:
            # Unknown connection type - return only universal nodes (safe)
            return cls.UNIVERSAL_NODES.copy()

    @classmethod
    def get_node_metadata(cls, connection_type: str) -> Dict[str, Dict[str, Any]]:
        """
        Get metadata for all nodes including availability and warnings.

        Args:
            connection_type: 'official' or 'qrcode'

        Returns:
            Dict with node_type as key and metadata as value
        """
        metadata = {}

        for node_type in cls.ALL_NODES:
            available = node_type in cls.get_available_nodes(connection_type)

            # Determine status
            if not available:
                status = "unavailable"
                warning = "Este node não está disponível para conexões Evolution API"
            elif node_type in cls.EXPERIMENTAL_NODES and connection_type == "qrcode":
                status = "experimental"
                warning = (
                    "Este node é experimental na Evolution API. "
                    "Funcionalidade pode ser limitada."
                )
            else:
                status = "available"
                warning = None

            metadata[node_type] = {
                "available": available,
                "status": status,
                "warning": warning,
                "connection_type": connection_type,
            }

        return metadata

    @classmethod
    def is_node_available(
        cls,
        node_type: str,
        connection_type: str
    ) -> bool:
        """
        Check if a specific node type is available for a connection type.

        Args:
            node_type: Type of node (e.g., 'whatsapp_template')
            connection_type: 'official' or 'qrcode'

        Returns:
            True if node is available, False otherwise
        """
        available_nodes = cls.get_available_nodes(connection_type)
        return node_type in available_nodes

    @classmethod
    def get_node_warning(
        cls,
        node_type: str,
        connection_type: str
    ) -> str | None:
        """
        Get warning message for a node type if applicable.

        Args:
            node_type: Type of node
            connection_type: 'official' or 'qrcode'

        Returns:
            Warning message or None
        """
        metadata = cls.get_node_metadata(connection_type)
        return metadata.get(node_type, {}).get("warning")


def get_available_nodes_for_number(connection_type: str) -> List[str]:
    """
    Helper function to get available nodes for a WhatsApp number.

    Args:
        connection_type: 'official' or 'qrcode'

    Returns:
        List of available node type strings
    """
    return NodeAvailability.get_available_nodes(connection_type)


def validate_node_compatibility(
    node_type: str,
    connection_type: str,
    raise_error: bool = True
) -> bool:
    """
    Validate if a node is compatible with a connection type.

    Args:
        node_type: Type of node to validate
        connection_type: 'official' or 'qrcode'
        raise_error: If True, raises ValueError on incompatibility

    Returns:
        True if compatible

    Raises:
        ValueError: If node is incompatible and raise_error=True
    """
    is_available = NodeAvailability.is_node_available(node_type, connection_type)
    
    if not is_available and raise_error:
        warning = NodeAvailability.get_node_warning(node_type, connection_type)
        raise ValueError(
            f"Node '{node_type}' não é compatível com '{connection_type}'. {warning}"
        )
    
    return is_available


def get_node_descriptions_for_ai() -> Dict[str, str]:
    """
    Get detailed descriptions of all node types for AI prompt generation.
    
    Returns:
        Dict mapping node_type to description string
    """
    return {
        "start": "Início do flow - Ponto de entrada do fluxo conversacional",
        "message": "Envia mensagem ao usuário - Texto simples ou com variáveis",
        "question": "Faz pergunta e armazena resposta em variável - Captura input do usuário",
        "condition": "Ramificação baseada em condições - if/else logic",
        "end": "Finaliza flow - Encerra a conversa",
        "handoff": "Transfere para atendente humano - Passa para queue ou departamento",
        "delay": "Pausa temporizada - Aguarda X segundos antes de continuar",
        "jump": "Navega para outro node ou flow - Redirecionamento",
        "action": "Executa ações (salvar contato, enviar email, webhook) - Side effects",
        "api_call": "Chama APIs externas - Integração com sistemas externos",
        "ai_prompt": "Usa IA para processar informações - GPT/Claude para análise",
        "database_query": "Consulta banco de dados - SELECT/INSERT/UPDATE queries",
        "script": "Executa código JavaScript - Lógica customizada",
        "set_variable": "Define ou atualiza variável - Manipulação de dados",
        "random": "Escolhe caminho aleatório - A/B testing ou variação",
        "datetime": "Trabalha com datas/horas - Agendamentos e validações temporais",
        "analytics": "Registra evento analítico - Tracking e métricas",
        "whatsapp_template": "Envia template aprovado pelo WhatsApp - Apenas Official API",
        "interactive_buttons": "Botões interativos WhatsApp (até 3 opções) - Apenas Official API",
        "interactive_list": "Lista interativa WhatsApp (4+ opções) - Apenas Official API",
    }
